rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,663 @@
+
+Dictionaries are dynamically-sized homogenous collections mapping _keys_ to _values_, where each key is unique in the instance. The key type is always a string (`Name` type). The `DictType` defines the value type.
+
+```rust
+#[derive(Component)]
+struct DictType {
+    value_type: AnyType,
+}
+```
+
+Dictionary values store their (key, value) pairs contiguously in the byte storage (see [Game objects](#game-objects)), key first followed by value, without padding.
+
+#### Structs
+
+A _struct_ is an heterogenous aggregation of other types (like a `struct` in Rust). The `StructType` component contains the _properties_ of the struct.
+
+```rust
+#[derive(Component)]
+struct StructType {
+    properties: Vec<Property>, // sorted by `byte_offset`
+}
+```
+
+Properties themselves are defined by a name and the type of their value. The type also stores the byte offset of that property from the beginning of a struct instance, which allows packing property values for a struct into a single byte stream, and enables various optimizations regarding diff'ing/patching and serialization. Modifying any of the `name`, `value_type`, or `byte_offset` constitute a type change, which mandates migrating all existing instances.
+
+```rust
+struct Property {
+    name: Name,
+    value_type: AnyType,
+    byte_offset: u32,
+    validate: Option<Box<dyn Validate>>,
+}
+```
+
+The _type layout_ refers to the collection of property offsets and types of an `StructType`, which together define:
+
+- the order in which the properties are listed in the struct type (the properties are sorted in increasing byte offset order), and any instance value stored in a `GameObject`.
+- the size in bytes of each property value, based on its type.
+- eventual gaps between properties, known as _padding bytes_, depending on their offsets and sizes.
+
+Unlike Rust, the Editor data model guarantees that **padding bytes are zeroed**. This property enables efficient instance comparison via `memcmp()`-like byte-level comparison, without the need to rely on individual properties inspection nor any equality operator (`PartialEq` and `Eq` in Rust). It also enables efficient copy and serialization via direct memory reads and writes, instead of relying on getter/setter methods.
+
+The `Property` struct contains also an optional validation function used to validate the value of a property for things like range and bounds check, nullability, _etc._ which is used both when the user (via the Editor UI) attempts to set a value, and when a value is set automatically (_e.g._ during deserialization).
+
+```rust
+trait Validate {
+    /// Check if a value is valid for the property.
+    fn is_valid(&self, property: &Property, value: &[u8]) -> bool;
+
+    /// Try to assign a value to a property of an instance.
+    fn try_set(&mut self, instance: &mut GameObject, property: &Property, value: &[u8]) -> bool;
+}
+```
+
+### Game objects
+
+Instances of any type are represented by the `GameObject` component:

```

Comment:
> any type

Can we be a little more specific on that? Any type from the Game? or any type from the ones defined above?

Reply:
Added mention of "any Game type".